operator/docs/enhancements/zone_aware_replication.md

```diff
@@ -0,0 +1,358 @@
+
+  // Decode the /bind request
+  err := wh.decoder.DecodeRaw(request.Object, binding)
+  if err != nil {...}
+
+
+  if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+     // 1. Read the current Pod specification
+     key := client.ObjectKey{Namespace: binding.ObjectMeta.Namespace, Name: binding.ObjectMeta.Name}
+     pod := &v1.Pod{}
+     if err := wh.client.Get(ctx, key, pod); err != nil {...}
+     // get the topology keys from the pods where the pod.spec.topologySpreadConstraints.topologyKey is set
+     kv, err := getTopologyKeyValue(ctx, wh.client, binding.Target.Name, pod.spec.topologySpreadConstraints.topologyKey)
+     if err != nil {...}
+     if pod.Annotations == nil {
+        pod.Annotations = make(map[string]string)
+     }
+     // 2. Add topology keys to the Pod's annotations
+     for k, v := range kv {
+        if k == topologyKey {
+          pod.Annotations[k] = v
+        }
+     }
+     // 3. Update the Pod
+     return wh.client.Update(ctx, pod)
+  }); err != nil {...}
+}
+```
+Some reference impementations on how to use the a `pods/binding` sub-resource mutating webhook can be found here:
+* [elastic/cloud-on-k8s](https://github.com/elastic/cloud-on-k8s/issues/3933#issuecomment-950581345)
+* [kubemod/kubemod](https://github.com/kubemod/kubemod/commit/95c43bc7d532db05051a83df5ad97e36a0f5d7e2#diff-d63889fb133b18cb690d365c25f43bf0580f7a7aaa91ffa03b121b9b6b5f2d60)
+
+2. The second solution is to use the operator to modify the pods once they are scheduled. The operator checks if the `pod.spec.topologySpreadConstraints.topologyKey` is set. If yes, then the operator extracts the annotation(key and value) from the node, and sets it as an annotation for the pod by patching the pod. The zone value can then be accessed from within the container as an environment variable just like in solution 1 by adding the env variable in the container. For reference consider:
+* [elastic/cloud-on-k8s implementation](https://github.com/elastic/cloud-on-k8s/blob/85aa75b7a7765330475c0b92a0851939ac3d9578/pkg/controller/elasticsearch/driver/node_labels.go#L57) 
+* [elastic/cloud-on-k8s discussion](https://github.com/elastic/cloud-on-k8s/issues/3933#issuecomment-962897004)
+
+This is how the expected individual pod spec will look like after the webhook has added the topology key annotation in the pod:
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: ingester
+  annotations: 
+    topology.kubernetes.io/zone: zone-a 
+spec:
+  topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        ingester: pod
+  containers:
+  - name: ingester
+    env:
+    - name: ZONE
+        valueFrom:
+          fieldRef:
+            fieldPath: metadata.annotations['topology.kubernetes.io/zone']
+```
+
+The container can then be started by passing the configuration via the cli flag `instance_availability_zone: $ZONE`
```

Comment:
One thing I missed in the discussion of both solutions is that there's only one `instance_availability_zone` setting, even if the user decides to use more than one label for the topology.

In the end we need to reduce all the information to a single "topology key" that we pass into Loki. It might be as simple as concatenating the different label values in a defined way, don't know if Loki has a maximum size on the configuration value, though.

Reply:
This is a point to save for later but a good catch. I am not sure if the following applies only to AWS, but I hardly remember GCP zones include the name of the region in them:
```
NAME                           STATUS   ROLES                  AGE   VERSION           REGION      ZONE
ip-10-0-139-124.ec2.internal   Ready    worker                 80m   v1.25.4+18eadca   us-east-1   us-east-1a
ip-10-0-139-24.ec2.internal    Ready    control-plane,master   89m   v1.25.4+18eadca   us-east-1   us-east-1a
ip-10-0-146-149.ec2.internal   Ready    worker                 81m   v1.25.4+18eadca   us-east-1   us-east-1b
ip-10-0-149-216.ec2.internal   Ready    control-plane,master   89m   v1.25.4+18eadca   us-east-1   us-east-1b
ip-10-0-166-102.ec2.internal   Ready    control-plane,master   88m   v1.25.4+18eadca   us-east-1   us-east-1c
ip-10-0-170-152.ec2.internal   Ready    worker                 81m   v1.25.4+18eadca   us-east-1   us-east-1c
ip-10-0-186-239.ec2.internal   Ready    worker                 81m   v1.25.4+18eadca   us-east-1   us-east-1d
ip-10-0-197-101.ec2.internal   Ready    worker                 82m   v1.25.4+18eadca   us-east-1   us-east-1e
ip-10-0-211-56.ec2.internal    Ready    worker                 81m   v1.25.4+18eadca   us-east-1   us-east-1f
```

Thus if we allow the users to use `topology.kubernetes.io/region` with `topology.kubernetes.io/zone` we can skip the first if the above observation is matching across public clouds at least. Of course this does not apply on custom labels or `kubernetes.io/hostname` as topology keys.

Reply:
GCP nomenclature is almost the same as AWS ([docs](https://cloud.google.com/compute/docs/regions-zones/)); regions are prefixes of zones.

I was not thinking of "public cloud users" when writing the comment but people in other environments who have different sets of labels that do not follow this prefix approach. One example I can think of is people labeling their nodes with "floor" and "rack" where rack is a number that can repeat for each of the floors. In that case the user might opt for using both for the topology which means for us we need to combine the values.

If I am honest I would put as little assumptions as possible into what the values of the different topology keys are and how they are related.